src/main/scala/beam/agentsim/infrastructure/ChargingNetworkManager.scala

```diff
@@ -0,0 +1,477 @@
+    }
+    chargingNetwork.processWaitingLine(station).foreach(handleStartCharging(tick, _))
+  }
+
+  /**
+    * Refuel the vehicle using last charging session and collect the corresponding load
+    * @param chargingVehicle vehicle charging information
+    */
+  private def handleRefueling(chargingVehicle: ChargingVehicle): Unit = {
+    chargingVehicle.latestChargingCycle.foreach { cycle =>
+      val vehicle = chargingVehicle.vehicle
+      vehicle.addFuel(cycle.energy)
+      collectObservedLoadInKW(chargingVehicle, cycle)
+      log.debug(s"Charging vehicle $vehicle. Stall ${vehicle.stall}. Provided energy of = ${cycle.energy} J")
+    }
+  }
+
+  /**
+    * process the event ChargingPlugInEvent
+    * @param currentTick current time
+    * @param chargingVehicle vehicle charging information
+    */
+  private def handleStartChargingHelper(currentTick: Int, chargingVehicle: ChargingVehicle): Unit = {
+    import chargingVehicle._
+    val chargingPlugInEvent = new ChargingPlugInEvent(
+      tick = currentTick,
+      stall = stall,
+      locationWGS = geo.utm2Wgs(stall.locationUTM),
+      vehId = vehicle.id,
+      primaryFuelLevel = vehicle.primaryFuelLevelInJoules,
+      secondaryFuelLevel = Some(vehicle.secondaryFuelLevelInJoules)
+    )
+    log.debug(s"ChargingPlugInEvent: $chargingPlugInEvent")
+    beamServices.matsimServices.getEvents.processEvent(chargingPlugInEvent)
+  }
+
+  /**
+    * Calculates the duration of the refuel session, the provided energy and throws corresponding events
+    * @param currentTick current time
+    * @param chargingVehicle vehicle charging information
+    */
+  def handleEndChargingHelper(currentTick: Int, chargingVehicle: ChargingVehicle): Unit = {
+    import chargingVehicle._
+    log.debug(
+      s"Vehicle $vehicle was disconnected at time {} with {} J delivered during {} sec",
+      currentTick,
+      chargingVehicle.computeSessionEnergy,
+      chargingVehicle.computeSessionDuration
+    )
+    // Refuel Session
+    val refuelSessionEvent = new RefuelSessionEvent(
+      currentTick,
+      stall.copy(locationUTM = geo.utm2Wgs(stall.locationUTM)),
+      chargingVehicle.computeSessionEnergy,
+      vehicle.primaryFuelLevelInJoules - chargingVehicle.computeSessionEnergy,
+      chargingVehicle.computeSessionDuration,
+      vehicle.id,
+      vehicle.beamVehicleType
+    )
+    log.debug(s"RefuelSessionEvent: $refuelSessionEvent")
+    beamServices.matsimServices.getEvents.processEvent(refuelSessionEvent)
+    val chargingPlugOutEvent: ChargingPlugOutEvent = new ChargingPlugOutEvent(
+      currentTick,
+      stall.copy(locationUTM = geo.utm2Wgs(stall.locationUTM)),
+      vehicle.id,
+      vehicle.primaryFuelLevelInJoules,
+      Some(vehicle.secondaryFuelLevelInJoules)
+    )
+    log.debug(s"ChargingPlugOutEvent: $chargingPlugOutEvent")
+    beamServices.matsimServices.getEvents.processEvent(chargingPlugOutEvent)
```

Comment:
Should this be done here? Shouldn't it be done alongside the actual plugout?

Reply:
it is